pkg/kallsyms/kallsyms.go

```diff
@@ -0,0 +1,209 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package kallsyms provides functions to read /proc/kallsyms.
+package kallsyms
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/cilium/ebpf"
+)
+
+type KAllSyms struct {
+	// symbols is a slice of kernel symbols. Order is preserved.
+	symbols []kernelSymbol
+
+	// symbolsMap is a map of kernel symbols. Provides fast lookup.
+	symbolsMap map[string]uint64
+}
+
+type kernelSymbol struct {
+	addr uint64
+	name string
+}
+
+// NewKAllSyms reads /proc/kallsyms and returns a KAllSyms.
+func NewKAllSyms() (*KAllSyms, error) {
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return NewKAllSymsFromReader(file)
+}
+
+// NewKAllSymsFromReader reads a kallsyms file from the given reader and returns
+// a KAllSyms.
+func NewKAllSymsFromReader(reader io.Reader) (*KAllSyms, error) {
+	symbols := []kernelSymbol{}
+	symbolsMap := map[string]uint64{}
+
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		addr, err := strconv.ParseUint(fields[0], 16, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		// The kernel function is the third field in /proc/kallsyms line:
+		// 0000000000000000 t acpi_video_unregister_backlight      [video]
+		// First is the symbol address and second is described in man nm.
+		symbols = append(symbols, kernelSymbol{
+			addr: addr,
+			name: fields[2],
+		})
+		symbolsMap[fields[2]] = addr
+	}
+
+	err := scanner.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	return &KAllSyms{
+		symbols:    symbols,
+		symbolsMap: symbolsMap,
+	}, nil
+}
+
+// LookupByInstructionPointer tries to find the kernel symbol corresponding to
+// the given instruction pointer.
+// For example, if instruction pointer is 0x1004 and there is a symbol which
+// address is 0x1000, this function will return the name of this symbol.
+// If no symbol is found, it returns "[unknown]".
+func (k *KAllSyms) LookupByInstructionPointer(ip uint64) string {
+	// Go translation of iovisor/bcc ksyms__map_addr():
+	// https://github.com/iovisor/bcc/blob/c65446b765c9f7df7e357ee9343192de8419234a/libbpf-tools/trace_helpers.c#L149
+	end := len(k.symbols) - 1
+	var addr uint64
+	start := 0
+
+	for start < end {
+		mid := start + (end-start+1)/2
+
+		addr = k.symbols[mid].addr
+
+		if addr <= ip {
+			start = mid
+		} else {
+			end = mid - 1
+		}
+	}
+
+	if start == end && k.symbols[start].addr <= addr {
```

Comment:
I know you didn't implement this, but I should check `k.symbols[start].addr <= ip` instead as done in BCC. ip here is called addr and I think it caused confusion when doing the porting. 

With the current check, a ksysall like 
```
		"ffffffff00000000 A fixed_percpu_data",
		"ffffffffb4231f40 D bpf_prog_fops",
		"ffffffffb43723e0 d socket_file_ops",
```

would return `fixed_percpu_data` for `LookupByInstructionPointer(0)` instead of `unknown`.